### PR TITLE
Fix null ref exception in Listener

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Fix null ref exception in listener (https://github.com/Azure/azure-sdk-for-net/issues/21359)
- Pass cancellation token to SB methods in listener
- Add checks for logged errors in tests
- Remove unnecessary ThrowIfCancellationRequested from processor Complete method

### CODE REVIEW

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs [hunk 1]
```diff
@@ -343,13 +361,14 @@ private async Task TestSingleDrainMode<T>(bool sendToQueue)
...
                 await jobHost.StopAsync();
+                AssertNoLoggedErrors(host);
             }
         }
 
         private static Action<IHostBuilder> BuildDrainHost<T>()
         {
             return builder =>
-                builder.ConfigureDefaultTestHost<T>(b =>
+                builder.ConfigureWebJobs(b =>
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs:371]: This was necessary as ConfigureDefaultTestHost added the logging provider, and BuildHost does as well so there were two logging providers and the GetTestLoggerProvider call was failing in the Drain tests.

**Review by jsquire — APPROVED** [review]

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/WebJobsServiceBusTestBase.cs [hunk 2]
```diff
@@ -215,5 +217,12 @@ internal async Task WriteTopicMessage(string message, string sessionId = null)
             }
             await sender.SendMessageAsync(messageObj);
         }
+
+        internal static void AssertNoLoggedErrors(IHost host)
```
Comment (pakrym) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/WebJobsServiceBusTestBase.cs:221]: Wonder if we can do this automatically on StopAsync.

Can we add an https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-5.0 implementation that does this as part of the shutdown?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs [hunk 3]
```diff
@@ -196,13 +196,28 @@ public async Task TestBatch_JsonPoco()
...
+            }
+        }
+
+        [Test]
+        public async Task TestBatch_NoMessages()
+        {
+            var (jobHost, host) = BuildHost<ServiceBusMultipleMessagesTestJob_NoMessagesExpected>();
+            using (jobHost)
+            {
+                // wait for 20 seconds to let the listener run for a while
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs:216]: can we tune some parameters so we don't need to wait 20 sec?

[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs:216]: What code-path executed the ServiceBusMultipleMessagesTestJob_NoMessagesExpected before? Why does it take 20 sec to be hit?

[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusEndToEndTests.cs:216]: This is a new test. I was being conservative with 20 seconds to account for the host startup.

**Pull request merged**